public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx

```diff
@@ -0,0 +1,526 @@
+  onChange: onChange;
+  theme: GrafanaTheme;
+  hide?: () => void;
+}
+
+interface BrowserState {
+  labels: SelectableLabel[];
+  searchTerm: string;
+}
+
+interface FacettableValue {
+  name: string;
+  // facetted?: boolean; // True if value is possible given the selector
+  selected?: boolean;
+}
+
+export interface SelectableLabel {
+  name: string;
+  selected?: boolean;
+  loading?: boolean;
+  values?: FacettableValue[];
+  hidden?: boolean;
+  facets?: number;
+}
+
+export function buildSelector(labels: SelectableLabel[]): string {
+  const selectedLabels = [];
+  for (const label of labels) {
+    if (label.selected && label.values && label.values.length > 0) {
+      const selectedValues = label.values.filter(value => value.selected).map(value => value.name);
+      if (selectedValues.length > 1) {
+        selectedLabels.push(`${label.name}=~"${selectedValues.join('|')}"`);
+      } else if (selectedValues.length === 1) {
+        selectedLabels.push(`${label.name}="${selectedValues[0]}"`);
+      }
+    }
+  }
+  return ['{', selectedLabels.join(','), '}'].join('');
+}
+
+export function facetLabels(
+  labels: SelectableLabel[],
+  possibleLabels: Record<string, string[]>,
+  lastFacetted?: string
+): SelectableLabel[] {
+  return labels.map(label => {
+    const possibleValues = possibleLabels[label.name];
+    if (possibleValues) {
+      let existingValues: FacettableValue[];
+      if (label.name === lastFacetted && label.values) {
+        // Facetting this label, show all values
+        existingValues = label.values;
+      } else {
+        // Keep selection in other facets
+        const selectedValues: Set<string> = new Set(
+          label.values?.filter(value => value.selected).map(value => value.name) || []
+        );
+        // Values for this label have not been requested yet, let's use the facetted ones as the initial values
+        existingValues = possibleValues.map(value => ({ name: value, selected: selectedValues.has(value) }));
+      }
+      return { ...label, loading: false, values: existingValues, facets: existingValues.length };
+    }
+
+    // Label is facetted out, hide all values
+    return { ...label, loading: false, hidden: !possibleValues, values: undefined, facets: 0 };
+  });
+}
+
+const getStyles = stylesFactory((theme: GrafanaTheme) => {
+  const { white, black, dark1, dark2, dark7, gray1, gray3, gray5, gray7 } = theme.palette;
+  const lightThemeStyles = {
+    linkColor: dark2,
+    linkColorHover: theme.colors.link,
+    wrapperBg: gray7,
+    wrapperShadow: gray3,
+    itemColor: black,
+    groupLabelColor: gray1,
+    itemBgHover: gray5,
+    headerBg: white,
+    headerSeparator: white,
+  };
+  const darkThemeStyles = {
+    linkColor: theme.colors.text,
+    linkColorHover: white,
+    wrapperBg: dark2,
+    wrapperShadow: black,
+    itemColor: white,
+    groupLabelColor: theme.colors.textWeak,
+    itemBgHover: dark7,
+    headerBg: dark1,
+    headerSeparator: dark7,
+  };
```

Comment:
**Maybe next iteration/during the beta**:
There are styles that are not used (linkColor, linkColorHover,...). This should be cleaned up and we should have here only styles that are used.